File: internal/pipeline/pipeline.go
// Package pipeline defines what a composed platform *is*: a list of module
// instances and the connections between them. A pipeline.json plus the
// prebuilt frontend and this binary is the entire exported product, which is
// why nothing here is specific to any one module.
package pipeline

import (
	"encoding/json"
	"fmt"
	"io"
	"regexp"
	"strings"

	"github.com/MaastrichtU-BISS/legal-blocks/internal/manifest"
)

// Node is one instance of a module in a pipeline. Two annotate steps in the
// same pipeline are two Nodes sharing a Module.
type Node struct {
	ID     string         `json:"id"`
	Module string         `json:"module"`
	Label  string         `json:"label"`
	Config map[string]any `json:"config,omitempty"`
}

// Endpoint identifies one port on one node.
type Endpoint struct {
	Node string `json:"node"`
	Port string `json:"port"`
}

func (e Endpoint) String() string { return e.Node + "." + e.Port }

// Edge connects an output port to an input port.
type Edge struct {
	From Endpoint `json:"from"`
	To   Endpoint `json:"to"`
}

// Pipeline is the exported platform's definition.
type Pipeline struct {
	Version int    `json:"version"`
	Name    string `json:"name"`
	// Mode says where this platform's data lives. It is a property of the
	// platform rather than of any module: the same annotation component is a
	// step in a durable multi-annotator task or a one-off pass over search
	// results, depending only on which source the host builds for it.
	//
	// Empty means persistent, so a pipeline written before modes existed keeps
	// the behaviour it had.
	Mode  manifest.Mode `json:"mode,omitempty"`
	Nodes []Node        `json:"nodes"`
	Edges []Edge        `json:"edges"`
}

// StorageMode returns the pipeline's mode, defaulting to persistent.
func (p *Pipeline) StorageMode() manifest.Mode {
	if p.Mode == "" {
		return manifest.ModePersistent
	}
	return p.Mode
}

var idPattern = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)

// Parse reads and validates a pipeline against a registry.
func Parse(r io.Reader, reg *manifest.Registry) (*Pipeline, error) {
	var p Pipeline
	if err := json.NewDecoder(r).Decode(&p); err != nil {
		return nil, fmt.Errorf("parsing pipeline: %w", err)
	}
	if err := p.Validate(reg); err != nil {
		return nil, err
	}
	return &p, nil
}

// Validate checks that every node names a known module, every edge connects
// ports that exist and carry compatible types, and every required input is
// connected. This is the same check the composer applies when the user draws
// a connection, so an exported pipeline cannot be one the composer would have
// rejected.
func (p *Pipeline) Validate(reg *manifest.Registry) error {
	if len(p.Nodes) == 0 {
		return fmt.Errorf("pipeline has no nodes")
	}

	mode := p.StorageMode()
	if mode != manifest.ModeEphemeral && mode != manifest.ModePersistent {
		return fmt.Errorf("unknown storage mode %q", mode)
	}

	byID := map[string]Node{}
	for _, n := range p.Nodes {
		if !idPattern.MatchString(n.ID) {
			return fmt.Errorf("node id %q must be alphanumeric, dash or underscore", n.ID)
		}
		if _, dup := byID[n.ID]; dup {
			return fmt.Errorf("duplicate node id %q", n.ID)
		}
		m, ok := reg.Modules[n.Module]
		if !ok {
			return fmt.Errorf("node %q references unknown module %q", n.ID, n.Module)
		}
		// A module that cannot work without stored resources has no meaning in
		// a platform that stores nothing, and vice versa. Catching it here
		// means an export cannot promise a screen that will not function.
		if !m.SupportsMode(mode) {
			return fmt.Errorf("module %q does not work in %s mode", n.Module, mode)
		}
		byID[n.ID] = n
	}

	// Every edge must land on ports that exist and carry compatible types.
	connected := map[string]bool{} // "node.port" of satisfied inputs
	for _, e := range p.Edges {
		fromNode, ok := byID[e.From.Node]
		if !ok {
			return fmt.Errorf("edge from unknown node %q", e.From.Node)
		}
		toNode, ok := byID[e.To.Node]
		if !ok {
			return fmt.Errorf("edge to unknown node %q", e.To.Node)
		}
		out, ok := findPort(reg.Modules[fromNode.Module].Outputs, e.From.Port)
		if !ok {
			return fmt.Errorf("module %q has no output port %q", fromNode.Module, e.From.Port)
		}
		in, ok := findPort(reg.Modules[toNode.Module].Inputs, e.To.Port)
		if !ok {
			return fmt.Errorf("module %q has no input port %q", toNode.Module, e.To.Port)
		}
		if !reg.CanConnect(out.Type, in.Type) {
			return fmt.Errorf("cannot connect %s (%s) to %s (%s): %s",
				e.From, out.Type, e.To, in.Type, whyNot(out.Type, in.Type))
		}
		if connected[e.To.String()] {
			return fmt.Errorf("input %s is connected more than once", e.To)
		}
		connected[e.To.String()] = true
	}

	// Required inputs are only required in a session platform.
	//
	// That is where an edge is how data reaches a step: a search feeds a
	// viewer, an upload feeds an annotator, and a step with nothing connected
	// has nothing to work on. With storage none of that is true. Documents
	// become datasets, a task names the dataset and labelset it uses, and the
	// annotate step is opened against a task somebody chose — so the corpus
	// arrives from the workspace, not from whatever happens to be upstream.
	//
	// Insisting on an edge there would mean drawing one that lies: connecting
	// upload to annotate would say "these documents" when the real answer is
	// "whichever dataset the task names". Edges that are present are still
	// type-checked above; they just stop being compulsory.
	if p.StorageMode() == manifest.ModeEphemeral {
		for _, n := range p.Nodes {
			for _, in := range reg.Modules[n.Module].Inputs {
				if in.Required && !connected[n.ID+"."+in.Name] {
					return fmt.Errorf("node %q has no connection for required input %q", n.ID, in.Name)
				}
			}
		}
	}

	return p.checkAcyclic()
}

// checkAcyclic rejects cycles. The composer only builds linear chains today,
// but the model is a graph and an exported pipeline.json can be hand-edited,
// so the runtime must not be able to loop forever resolving inputs.
func (p *Pipeline) checkAcyclic() error {
	deps := map[string][]string{}
	for _, e := range p.Edges {
		deps[e.To.Node] = append(deps[e.To.Node], e.From.Node)
	}
	const (
		visiting = 1
		done     = 2
	)
	state := map[string]int{}
	var walk func(string) error
	walk = func(id string) error {
		switch state[id] {
		case done:
			return nil
		case visiting:
			return fmt.Errorf("pipeline contains a cycle through node %q", id)
		}
		state[id] = visiting
		for _, dep := range deps[id] {
			if err := walk(dep); err != nil {
				return err
			}
		}
		state[id] = done
		return nil
	}
	for _, n := range p.Nodes {
		if err := walk(n.ID); err != nil {
			return err
		}
	}
	return nil
}

// Order returns node IDs in dependency order — every node after the nodes
// feeding it. The runtime uses this to lay out the step navigation.
func (p *Pipeline) Order() []string {
	deps := map[string][]string{}
	for _, e := range p.Edges {
		deps[e.To.Node] = append(deps[e.To.Node], e.From.Node)
	}
	var out []string
	seen := map[string]bool{}
	var walk func(string)
	walk = func(id string) {
		if seen[id] {
			return
		}
		seen[id] = true
		for _, dep := range deps[id] {
			walk(dep)
		}
		out = append(out, id)
	}
	for _, n := range p.Nodes {
		walk(n.ID)
	}
	return out
}

// ServiceIDs lists the Go services the pipeline's modules require, so the host
// mounts only what is actually used.
func (p *Pipeline) ServiceIDs(reg *manifest.Registry) []string {
	seen := map[string]bool{}
	var out []string
	for _, n := range p.Nodes {
		for _, s := range reg.Modules[n.Module].Services {
			if !seen[s] {
				seen[s] = true
				out = append(out, s)
			}
		}
	}
	return out
}

// whyNot explains a refused connection. The generic answer is true but
// unhelpful — someone wiring a search into an annotation step has a reasonable
// idea and is missing a piece, and saying which piece is the difference
// between a dead end and a next step.
func whyNot(from, to string) string {
	if from == "document-set@1" && to == "corpus@1" {
		return "search results are cases, not documents to work on. Getting their text " +
			"means fetching each judgment, which is a preprocessing step rather than " +
			"something that can happen on this connection"
	}
	return "no adapter declared"
}

func findPort(ports []manifest.Port, name string) (manifest.Port, bool) {
	for _, p := range ports {
		if strings.EqualFold(p.Name, name) {
			return p, true
		}
	}
	return manifest.Port{}, false
}

File: packages/manifest/src/pipeline.ts
// What a composed platform *is*: an ordered list of module instances. A
// pipeline.json plus the platform image is the entire exported product, which
// is why nothing here is specific to any one module.
//
// The order is the wiring. A pipeline runs its steps front to back and nobody
// can skip one, so step 2 reads what step 1 produced — there is nothing for a
// connection to say that the position does not already say. This used to be a
// graph with an explicit edge list, and the edges only ever described the
// chain the composer had already laid out.

import type { Kind } from "./kinds.js";
import type { Manifest, Port, Registry } from "./manifest.js";
import { canConnect, supportsKind } from "./manifest.js";

/**
 * One instance of a module in a pipeline. Two annotate steps in the same
 * pipeline are two nodes sharing a module.
 */
export interface Node {
  id: string;
  module: string;
  label?: string;
  config?: Record<string, unknown>;
}

/** The exported platform's definition. */
export interface Pipeline {
  version: number;
  name: string;
  /**
   * What this is: a pipeline that runs start to finish and keeps nothing, or a
   * workspace people come back to. A property of the export rather than of any
   * module — the same annotation component is a step in a one-off pass over
   * search results or a durable multi-annotator task, depending only on which
   * source the platform builds for it.
   *
   * Absent means workspace, so a file written before this field existed keeps
   * the behaviour it had.
   */
  kind?: Kind;
  /** The steps, in the order they run. */
  nodes: Node[];
}

/** What this is, defaulting to a workspace. */
export function exportKind(p: Pipeline): Kind {
  return p.kind ?? "workspace";
}

const ID_PATTERN = /^[a-zA-Z0-9_-]+$/;

/**
 * Reads and validates a pipeline against a registry.
 *
 * Throws with the message the user should see: these are read by whoever is
 * composing, and "cannot connect a to b" is the whole feedback loop.
 */
export function parsePipeline(raw: unknown, reg: Registry): Pipeline {
  const p = migrate((typeof raw === "string" ? JSON.parse(raw) : raw) as Pipeline);
  validatePipeline(p, reg);
  return p;
}

/**
 * Brings a pipeline written when this was a graph up to the ordered list.
 *
 * An export's compose file invites upgrading in place by changing an image
 * tag, so a pipeline.json written by an older composer will meet a newer
 * platform. Back then the runtime laid steps out in dependency order rather
 * than array order, and the two could differ — so reading such a file means
 * putting the nodes in the order its edges implied, then forgetting them.
 *
 * Anything that is already a list passes through untouched.
 */
function migrate(p: Pipeline): Pipeline {
  const legacy = (p as { edges?: { from?: { node?: string }; to?: { node?: string } }[] }).edges;
  if (!Array.isArray(legacy) || legacy.length === 0 || !Array.isArray(p.nodes)) return p;

  const deps = new Map<string, string[]>();
  for (const e of legacy) {
    const to = e?.to?.node;
    const from = e?.from?.node;
    if (typeof to === "string" && typeof from === "string") {
      deps.set(to, [...(deps.get(to) ?? []), from]);
    }
  }

  const byId = new Map(p.nodes.map((n) => [n.id, n]));
  const ordered: Node[] = [];
  const seen = new Set<string>();
  const walk = (id: string): void => {
    // A cycle in a hand-edited legacy file stops here rather than looping: the
    // result is still a list, which is the only shape that can run now.
    if (seen.has(id)) return;
    seen.add(id);
    for (const dep of deps.get(id) ?? []) walk(dep);
    const node = byId.get(id);
    if (node) ordered.push(node);
  };
  for (const n of p.nodes) walk(n.id);

  const { edges: _dropped, ...rest } = p as Pipeline & { edges?: unknown };
  return { ...rest, nodes: ordered };
}

/**
 * Checks that every node names a known module, and that in a pipeline each
 * step's required inputs are satisfied by the step in front of it.
 *
 * This is the same check the composer applies when deciding what may be added
 * next, so an exported pipeline cannot be one the composer would have refused
 * to build.
 */
export function validatePipeline(p: Pipeline, reg: Registry): void {
  if (!p || !Array.isArray(p.nodes) || p.nodes.length === 0) {
    throw new Error("pipeline has no nodes");
  }

  const kind = exportKind(p);
  if (kind !== "pipeline" && kind !== "workspace") {
    throw new Error(`unknown kind "${kind}" — expected pipeline or workspace`);
  }

  const byId = new Map<string, Node>();
  for (const n of p.nodes) {
    if (!ID_PATTERN.test(n.id ?? "")) {
      throw new Error(`node id "${n.id}" must be alphanumeric, dash or underscore`);
    }
    if (byId.has(n.id)) {
      throw new Error(`duplicate node id "${n.id}"`);
    }
    const m = reg.modules[n.module];
    if (!m) {
      throw new Error(`node "${n.id}" references unknown module "${n.module}"`);
    }
    // A module that needs stored resources has no meaning in a pipeline, and
    // vice versa. Catching it here means an export cannot promise a screen
    // that will not function.
    if (!supportsKind(m, kind)) {
      throw new Error(`module "${n.module}" does not belong in a ${kind}`);
    }
    byId.set(n.id, n);
  }

  // Required inputs are only required in a pipeline.
  //
  // That is where the order is how data reaches a step: a search feeds a
  // viewer, an upload feeds an annotator, and a first step that needs input
  // has nothing to work on. In a workspace none of that is true. Documents
  // become datasets, a task names the dataset and labelset it uses, and the
  // annotate tool is opened against a task somebody chose — so the corpus
  // arrives from the workspace, not from the step before.
  if (kind === "pipeline") {
    p.nodes.forEach((n, i) => {
      for (const input of reg.modules[n.module]?.inputs ?? []) {
        if (!input.required) continue;

        const previous = p.nodes[i - 1];
        if (!previous) {
          throw new Error(
            `step "${n.id}" needs ${input.type} to work on, but it is the first step — ` +
              `put a step that produces ${input.type} in front of it`,
          );
        }
        const supplied = supplies(reg, previous, input);
        if (!supplied) {
          const produced = reg.modules[previous.module]?.outputs?.[0]?.type;
          throw new Error(
            `step "${n.id}" needs ${input.type}, but the step before it ("${previous.id}") ` +
              `produces ${produced ?? "nothing"}: ${whyNot(produced ?? "", input.type)}`,
          );
        }
      }
    });
  }
}

/**
 * The port on `previous` that feeds `input`, if any.
 *
 * The first compatible output wins. A module with several outputs whose types
 * a downstream step could take either of is not something the registry has,
 * and picking the first keeps the rule sayable: a step reads what the step
 * before it produces.
 */
function supplies(reg: Registry, previous: Node, input: Port): Port | undefined {
  return reg.modules[previous.module]?.outputs?.find((out) => canConnect(reg, out.type, input.type));
}

/** Node ids in the order they run, which is the order they are written in. */
export function order(p: Pipeline): string[] {
  return p.nodes.map((n) => n.id);
}

/**
 * The services the pipeline's modules require, so the platform mounts only
 * what is actually used.
 */
export function serviceIds(p: Pipeline, reg: Registry): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const n of p.nodes) {
    for (const s of reg.modules[n.module]?.services ?? []) {
      if (!seen.has(s)) {
        seen.add(s);
        out.push(s);
      }
    }
  }
  return out;
}

/**
 * Explains a refused connection.
 *
 * The generic answer is true but unhelpful — someone wiring a search into an
 * annotation step has a reasonable idea and is missing a piece, and saying
 * which piece is the difference between a dead end and a next step.
 */
function whyNot(from: string, to: string): string {
  if (from === "document-set@1" && to === "corpus@1") {
    return (
      "search results are cases, not documents to work on. Getting their text " +
      "means fetching each judgment, which is a preprocessing step rather than " +
      "something that can happen on this connection"
    );
  }
  return "no adapter declared";
}

/** The manifest for a node, or undefined if the registry does not have it. */
export function moduleOf(p: Pipeline, reg: Registry, nodeId: string): Manifest | undefined {
  const node = p.nodes.find((n) => n.id === nodeId);
  return node ? reg.modules[node.module] : undefined;
}
